File: VB/CustomBandedGridView/CustomView.vb
Imports DevExpress.Utils
Imports DevExpress.XtraGrid
Imports DevExpress.XtraGrid.Columns
Imports DevExpress.XtraGrid.Registrator
Imports DevExpress.XtraGrid.Views.BandedGrid
Imports DevExpress.XtraGrid.Views.Base
Imports DevExpress.XtraGrid.Views.Base.Handler
Imports DevExpress.XtraGrid.Views.Grid
Imports DevExpress.XtraGrid.Views.Grid.ViewInfo
Imports DevExpress.XtraGrid.Views.Printing
Imports DevExpress.XtraPrinting
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Linq
Imports System.Reflection
Imports System.Text
Imports System.Windows.Forms

Namespace dxExample
    Public Class CustomView
        Inherits GridView

        Public Const SViewName As String = "CustomView"
        Public Sub New()
            Me.New(Nothing)
        End Sub
        Public Sub New(ByVal grid As GridControl)
            MyBase.New(grid)
            HeaderHeightOnExport = 50
        End Sub
        Public Property HeaderHeightOnExport() As Integer
        Protected Overrides ReadOnly Property ViewName() As String
            Get
                Return SViewName
            End Get
        End Property

        Protected Overrides Function CreatePrintInfoInstance(ByVal args As PrintInfoArgs) As BaseViewPrintInfo
            Return New CustomPrintInfo(args)
        End Function
    End Class
    Public Class CustomPrintInfo
        Inherits GridViewPrintInfo

        Public Sub New(ByVal args As PrintInfoArgs)
            MyBase.New(args)
        End Sub

        Public Overrides Sub PrintHeader(ByVal graph As BrickGraphics)
            MyBase.PrintHeader(graph)
            If Not View.OptionsPrint.PrintHeader Then
                Return
            End If
            Dim indent As New Point(Me.Indent, HeaderY)
            Dim r As Rectangle = Rectangle.Empty
            Dim usePrintStyles As Boolean = View.OptionsPrint.UsePrintStyles
            SetDefaultBrickStyle(graph, Bricks("HeaderPanel"))

            Dim HeaderHeight As Integer = 50
            For Each col As PrintColumnInfo In Columns
                If Not usePrintStyles Then
                    Dim temp As New AppearanceObject()
                    AppearanceHelper.Combine(temp, New AppearanceObject() { col.Column.AppearanceHeader, View.Appearance.HeaderPanel, AppearancePrint.HeaderPanel })
                    SetDefaultBrickStyle(graph, Bricks.Create(temp, BorderSide.All, temp.BorderColor, 1))
                End If
                r = col.Bounds

                r = New Rectangle(r.Location, New Size(r.Width, HeaderHeight))
                r.Offset(indent)
                Dim mi As MethodInfo = GetType(GridColumn).GetMethod("GetTextCaptionForPrinting", BindingFlags.NonPublic Or BindingFlags.Instance)
                Dim caption As String = DirectCast(mi.Invoke(col.Column, New Object() { }), String) 'col.Column.GetTextCaptionForPrinting();
                If Not col.Column.OptionsColumn.ShowCaption Then
                    caption = String.Empty
                End If
                Dim itb As ITextBrick = DrawTextBrick(graph, caption, r)

                If caption.Contains(Environment.NewLine) Then
                    itb.Style.StringFormat = BrickStringFormat.Create(itb.Style.TextAlignment, True)
                End If
                itb.Style.StringFormat = itb.Style.StringFormat.ChangeFormatFlags(StringFormatFlags.DirectionVertical Or StringFormatFlags.DirectionRightToLeft)

                If AppearancePrint.HeaderPanel.TextOptions.WordWrap = WordWrap.NoWrap AndAlso View.OptionsPrint.UsePrintStyles Then
                    Using g As Graphics = Me.View.GridControl.CreateGraphics()
                        Dim s As SizeF = g.MeasureString(itb.Text, itb.Font, 1000, itb.StringFormat.Value)
                        If s.Width + 5 >= r.Width Then
                            itb.Text = ""
                            itb.TextValue = ""
                        End If
                    End Using
                End If
                Using g As Graphics = Me.View.GridControl.CreateGraphics()
                    Dim s As SizeF = g.MeasureString(itb.Text, itb.Font, 1000, itb.StringFormat.Value)

                    If s.Height + 5 >= r.Height Then
                        itb.Text = ""
                        itb.TextValue = ""
                    End If
                End Using
            Next col
        End Sub
    End Class
End Namespace

File: CS/CustomBandedGridView/CustomView.cs
using DevExpress.Utils;
using DevExpress.XtraGrid;
using DevExpress.XtraGrid.Columns;
using DevExpress.XtraGrid.Registrator;
using DevExpress.XtraGrid.Views.BandedGrid;
using DevExpress.XtraGrid.Views.Base;
using DevExpress.XtraGrid.Views.Base.Handler;
using DevExpress.XtraGrid.Views.Grid;
using DevExpress.XtraGrid.Views.Grid.ViewInfo;
using DevExpress.XtraGrid.Views.Printing;
using DevExpress.XtraPrinting;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Reflection;
using System.Text;
using System.Windows.Forms;

namespace dxExample {
    public class CustomView : GridView {
        public const string SViewName = "CustomView";
        public CustomView() : this(null) { }
        public CustomView(GridControl grid) : base(grid) {
            HeaderHeightOnExport = 50;
        }
        public int HeaderHeightOnExport { get; set; }
        protected override string ViewName { get { return SViewName; } }

        protected override BaseViewPrintInfo CreatePrintInfoInstance(PrintInfoArgs args) {
            return new CustomPrintInfo(args);
        }
    }
    public class CustomPrintInfo : GridViewPrintInfo {
        public CustomPrintInfo(PrintInfoArgs args) : base(args) { }

        public override void PrintHeader(BrickGraphics graph) {
            base.PrintHeader(graph);
            if(!View.OptionsPrint.PrintHeader) return;
            Point indent = new Point(Indent, HeaderY);
            Rectangle r = Rectangle.Empty;
            bool usePrintStyles = View.OptionsPrint.UsePrintStyles;
            SetDefaultBrickStyle(graph, Bricks["HeaderPanel"]);

            int HeaderHeight = 50;
            foreach(PrintColumnInfo col in Columns) {
                if(!usePrintStyles) {
                    AppearanceObject temp = new AppearanceObject();
                    AppearanceHelper.Combine(temp, new AppearanceObject[] { col.Column.AppearanceHeader, View.Appearance.HeaderPanel, AppearancePrint.HeaderPanel });
                    SetDefaultBrickStyle(graph, Bricks.Create(temp, BorderSide.All, temp.BorderColor, 1));
                }
                r = col.Bounds;

                r = new Rectangle(r.Location, new Size(r.Width, HeaderHeight));
                r.Offset(indent);
                MethodInfo mi = typeof(GridColumn).GetMethod("GetTextCaptionForPrinting", BindingFlags.NonPublic | BindingFlags.Instance);
                string caption = (string)mi.Invoke(col.Column, new object[] { });//col.Column.GetTextCaptionForPrinting();
                if(!col.Column.OptionsColumn.ShowCaption) caption = string.Empty;
                ITextBrick itb = DrawTextBrick(graph, caption, r);

                if(caption.Contains(Environment.NewLine))
                    itb.Style.StringFormat = BrickStringFormat.Create(itb.Style.TextAlignment, true);
                itb.Style.StringFormat = itb.Style.StringFormat.ChangeFormatFlags(StringFormatFlags.DirectionVertical | StringFormatFlags.DirectionRightToLeft);

                if(AppearancePrint.HeaderPanel.TextOptions.WordWrap == WordWrap.NoWrap && View.OptionsPrint.UsePrintStyles) {
                    using(Graphics g = this.View.GridControl.CreateGraphics()) {
                        SizeF s = g.MeasureString(itb.Text, itb.Font, 1000, itb.StringFormat.Value);
                        if(s.Width + 5 >= r.Width) {
                            itb.Text = "";
                            itb.TextValue = "";
                        }
                    }
                }
                using(Graphics g = this.View.GridControl.CreateGraphics()) {
                    SizeF s = g.MeasureString(itb.Text, itb.Font, 1000, itb.StringFormat.Value);

                    if(s.Height + 5 >= r.Height) {
                        itb.Text = "";
                        itb.TextValue = "";
                    }
                }
            }
        }
    }
}
